pkg/dynamiccache/cache.go

```diff
@@ -0,0 +1,66 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package dynamiccache
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/open-policy-agent/gatekeeper/pkg/dynamiccache/internal"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+var log = logf.Log.WithName("object-cache")
+
+var defaultResyncTime = 10 * time.Hour
+
+// New initializes and returns a new Cache.
+func New(config *rest.Config, opts cache.Options) (cache.Cache, error) {
+	opts, err := defaultOpts(config, opts)
+	if err != nil {
+		return nil, err
+	}
+	im := internal.NewInformersMap(config, opts.Scheme, opts.Mapper, *opts.Resync, opts.Namespace)
+	return &dynamicInformerCache{InformersMap: im}, nil
+}
+
+func defaultOpts(config *rest.Config, opts cache.Options) (cache.Options, error) {
+	// Use the default Kubernetes Scheme if unset
+	if opts.Scheme == nil {
+		opts.Scheme = scheme.Scheme
+	}
+
+	// Construct a new Mapper if unset
+	if opts.Mapper == nil {
+		var err error
+		opts.Mapper, err = apiutil.NewDiscoveryRESTMapper(config)
```

Comment:
We should be sure to use the dynamic rest mapper here so that we can handle newly created resources.

Question: is it ever possible for the rest mapper to become stale (i.e. kind or resource name gets re-used)? I'm guessing not as the resource -> gvk mapping (and vice versa) should be stable, right?

Reply:
^ this code path was not being hit as the manager sets the dynamic rest mapper [here](https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/manager/manager.go#L359) and then passes it to the cache [here](https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/manager/manager.go#L238-L249).

But yes this is a bug waiting to happen, I'll change it.

Regarding your question, I believe the mapping is stable as well. What might change over time is version priority? Not sure this affects us.

Reply:
DynamicRESTMapper is now set explicitly.